### PR TITLE
feat: Add wasm support in `wrangler pages publish`

### PR DESCRIPTION
**Pull request by CarmenPopoviciu**
**What this PR solves:**
Currently it is not possible to import `wasm` modules in either Pages
Functions or Pages Advanced Mode projects.

This commit caries out work to address the aforementioned issue by
enabling `wasm` module imports in `wrangler pages publish`. As a result,
Pages users can now import their `wasm` modules withing their Functions
or `_worker.js` files, and `wrangler pages publish` will correctly
bundle everything and serve these "external" modules.

```
import hello from "./hello.wasm"

export async function onRequest() {
	const module = await WebAssembly.instantiate(hello);
	return new Response(module.exports.hello);
}
```

**How to test:**
Use the fixtures projects provided in this PR

- `pages-functions-wasm-app`

```
# build wrangler
npm run build

# cd into the fixture folder
cd fixtures/pages-functions-wasm-app

# publish project
npx wrangler pages publish public --experimental-worker-bundle=true
```

- `pages-workerjs-wasm-app`

```
# build wrangler
npm run build

# cd into the fixture folder
cd fixtures/pages-workerjs-wasm-app

# publish project
npx wrangler pages publish public --experimental-worker-bundle=true
```

Associated docs issues/PR:
N/A

Author has included the following, where applicable:

- [x] Tests
- [x] Changeset

Reviewer has performed the following, where applicable:

- [ ] Checked for inclusion of relevant tests
- [ ] Checked for inclusion of a relevant changeset
- [ ] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

Fixes # [insert issue number].

### CODE REVIEW

File: packages/wrangler/src/api/pages/publish.tsx
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/pages/publish.tsx:270]: not sure if there's a better way. happy to hear if there is 👀

File: packages/wrangler/src/api/pages/publish.tsx
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/pages/publish.tsx:276]: @GregBrimble if we do smth like this 👆, would we still need those 70 bytes thing we talked abt? Unless `undici` changes how they generate the response body from the formData
👉 https://github.com/nodejs/undici/blob/f02ab5b18a7e5cfcf0ffafaf8dc7a88b1f8430e3/lib/fetch/response.js#L156
👉 https://github.com/nodejs/undici/blob/f02ab5b18a7e5cfcf0ffafaf8dc7a88b1f8430e3/lib/fetch/response.js#L476
👉 https://github.com/nodejs/undici/blob/f02ab5b18a7e5cfcf0ffafaf8dc7a88b1f8430e3/lib/fetch/body.js#L31
👉 https://github.com/nodejs/undici/blob/f02ab5b18a7e5cfcf0ffafaf8dc7a88b1f8430e3/lib/fetch/body.js#L108

the first line of this body will contain just the undici-generated boundary. I would assume they are returning a pretty standard format, but I'm pretty new with this stuff, so I don't want to assume anything

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:14]: personally I think is one level of abstraction too many. `createWorkerBundleFormData` should prolly be enough. Whoever consumes this API can decide what they want to do with the formData, but maybe that's me, so I'm asking for a second opinion here :)

`createUploadWorkerBundleContents` is called twice in `publish`

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:63]: @GregBrimble regarding us stripping away those `bindings` from `metadata`... It seems that the `WorkerMetadata` interface is very specific in requiring those bindings to be present 👇 

```
export interface WorkerMetadata {
	main_module?: string;
	body_part?: string;
	compatibility_date?: string;
	compatibility_flags?: string[];
	usage_model?: "bundled" | "unbound";
	migrations?: CfDurableObjectMigrations;
	capnp_schema?: string;
	bindings: WorkerMetadataBinding[];
	keep_bindings?: WorkerMetadataBinding["type"][];
	logpush?: boolean;
}
```
I can still try to strip them from our `metadata` before we send them further off into the world ;). Unless we think we might very possible need them as well at some point?

File: packages/wrangler/src/pages/buildFunctions.ts
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/buildFunctions.ts:98]: @GregBrimble because you have more context than me here ... I know we said we leave plugins untouched for now. Do we maybe want to turn this feat on for plugin authors in a separate PR?

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/buildFunctions.ts:98]: discussed this IRL, and we concluded we'll park this for now. resolving...

File: packages/wrangler/src/api/pages/publish.tsx
Comment (GregBrimble) [inline packages/wrangler/src/api/pages/publish.tsx:251]:
```suggestion
		const enableBundling = (bundle ?? !noBundle) || experimentalWorkerBundle;
```

File: packages/wrangler/src/api/pages/publish.tsx
Comment (GregBrimble) [inline packages/wrangler/src/api/pages/publish.tsx:266]: We talked about this offline, but just for posterity: I think it would be _neater_ if this `buildRawWorker` and `buildFunctions` functions, which take an `outfile` prop, used that prop to write the bundle to, when `experimentalWorkerBundle` is true. I know we don't because these functions are also used by `wrangler pages dev`, but it feels like it should be `buildRawWorker`/`buildFunctions`'s responsibility to write to this file when they have all this information upfront. No big stress though. I'm happy to leave as-is for now. After we ship it, we can do a bit of cleanup when we get a second.

File: packages/wrangler/src/__tests__/pages/publish.test.ts
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages/publish.test.ts:2311]: Likewise, it would be nice if we could omit the empty `"bindings":[]`, but not critical and we can do it in cleanup if we still care enough.

File: packages/wrangler/src/__tests__/pages/publish.test.ts
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages/publish.test.ts:2293]: Pretty nice way to test this :)

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (GregBrimble) [inline packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:18]:
```suggestion
		workerBundle
```

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (GregBrimble) [inline packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:19]:
```suggestion
	const workerBundleFormData = createWorkerBundleFormData(workerBundle);
```

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (GregBrimble) [reply packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:14]: I think this is fine as you've got it :)

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (GregBrimble) [reply packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:63]: I think it would be nice if we could omit the empty `"bindings":[]`, but totally not critical and we can do it in cleanup if we still care enough.

**Review by GregBrimble — APPROVED** [review]

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:14]: yeah, you are right. this came very in handy for the `functions build` work, so I take my comment back 😅

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:63]: nah, let's do this now. You know how cleanup works 😋

Comment (CarmenPopoviciu) [reply packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:63]: ✅

File: packages/wrangler/src/api/pages/create-worker-bundle-contents.ts
Comment (GregBrimble) [inline packages/wrangler/src/api/pages/create-worker-bundle-contents.ts:26]:
```suggestion
	let metadata = JSON.parse(workerBundleFormData.get("metadata") as string);

	/**
	 * Pages doesn't need the metadata bindings returned by
	 * `createWorkerBundleFormData`. Let's strip them out and return only
	 * the contents we need
	 */
	workerBundleFormData.set("metadata", JSON.stringify({ main_module: metadata.main_module }));
```

**Pull request merged**
